File: 0684/sesión22 - WS - RESTFul/Back-End/ServidorRESTFul/SoftVenPersistance/src/main/java/pe/edu/pucp/softven/daoImp/DAOImplBase.java
package pe.edu.pucp.softven.daoImp;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import java.util.logging.Level;
import java.util.logging.Logger;
import pe.edu.pucp.softven.daoImp.util.Columna;
import pe.edu.pucp.softven.daoImp.util.Tipo_Operacion;
import pe.edu.pucp.softven.db.DBManager;

public abstract class DAOImplBase {

    protected String nombre_tabla;
    protected ArrayList<Columna> listaColumnas;
    protected Boolean retornarLlavePrimaria;
    protected Boolean mostrarSentenciaSQL;
    protected Connection conexion;
    protected CallableStatement statement;
    protected ResultSet resultSet;

    public DAOImplBase(String nombre_tabla) {
        this.nombre_tabla = nombre_tabla;
        this.retornarLlavePrimaria = false;
        this.mostrarSentenciaSQL = true;
        this.incluirListaDeColumnas();
    }

    private void incluirListaDeColumnas() {
        this.listaColumnas = new ArrayList<>();
        this.configurarListaDeColumnas();
    }

    protected abstract void configurarListaDeColumnas();

    protected void abrirConexion() {
        this.conexion = DBManager.getInstance().getConnection();
    }

    protected void cerrarConexion() throws SQLException {
        if (this.conexion != null) {
            this.conexion.close();
            this.conexion = null;
        }
    }

    protected void iniciarTransaccion() throws SQLException {
        this.abrirConexion();
        this.conexion.setAutoCommit(false);
    }

    protected void comitarTransaccion() throws SQLException {
        this.conexion.commit();
    }

    protected void rollbackTransaccion() throws SQLException {
        if (this.conexion != null) {
            this.conexion.rollback();
        }
    }

    protected void colocarSQLenStatement(String sql) throws SQLException {
        this.statement = this.conexion.prepareCall(sql);
    }

    protected Integer ejecutarModificacionEnBD() throws SQLException {
        if (this.mostrarSentenciaSQL) {
            Logger.getLogger(DAOImplBase.class.getName()).log(Level.INFO, this.statement.toString());
        }
        return this.statement.executeUpdate();
    }

    protected void ejecutarConsultaEnBD() throws SQLException {
        this.resultSet = this.statement.executeQuery();
    }

    protected Integer insertar() {
        return this.ejecuta_DML(Tipo_Operacion.INSERTAR);
    }

    protected Integer modificar() {
        return this.ejecuta_DML(Tipo_Operacion.MODIFICAR);
    }

    protected Integer eliminar() {
        return this.ejecuta_DML(Tipo_Operacion.ELIMINAR);
    }

    private Integer ejecuta_DML(Tipo_Operacion tipo_Operacion) {
        int resultado = 0;
        try {
            this.iniciarTransaccion();
            //this.conexion = DBManager.getInstance().getConnection();
            //this.conexion.setAutoCommit(false);
            String sql = null;
            switch (tipo_Operacion) {
                case Tipo_Operacion.INSERTAR ->
                    sql = this.generarSQLParaInsercion();
                case Tipo_Operacion.MODIFICAR ->
                    sql = this.generarSQLParaModificacion();
                case Tipo_Operacion.ELIMINAR ->
                    sql = this.generarSQLParaEliminacion();
            }
            this.colocarSQLenStatement(sql);
            switch (tipo_Operacion) {
                case Tipo_Operacion.INSERTAR ->
                    this.incluirValorDeParametrosParaInsercion();
                case Tipo_Operacion.MODIFICAR ->
                    this.incluirValorDeParametrosParaModificacion();
                case Tipo_Operacion.ELIMINAR ->
                    this.incluirValorDeParametrosParaEliminacion();
            }
            resultado = this.ejecutarModificacionEnBD();
            if (this.retornarLlavePrimaria && tipo_Operacion == Tipo_Operacion.INSERTAR) {
                resultado = this.retornarUltimoAutoGenerado();
            }
            this.comitarTransaccion();
        } catch (SQLException ex) {
            System.err.println("Error al intentar ejecutar consulta - " + tipo_Operacion.toString() + ": " + ex);
            try {
                this.rollbackTransaccion();
                //if (this.conexion != null) {
                //    this.conexion.rollback();
                //}
            } catch (SQLException ex1) {
                System.err.println("Error al hacer rollback - " + ex);
            }
        } finally {
            try {
                this.cerrarConexion();
                //if (this.conexion != null) {
                //    this.conexion.close();
                //}
            } catch (SQLException ex) {
                System.err.println("Error al cerrar la conexión - " + ex);
            }
        }
        return resultado;
    }

    protected String generarSQLParaInsercion() {
        //sentencia SQL a generar es similar a 
        //INSERT INTO INV_ALMACENES (NOMBRE, ALMACEN_CENTRAL) VALUES (?,?)
        String sql = "INSERT INTO ";
        sql = sql.concat(this.nombre_tabla);
        sql = sql.concat("(");
        String sql_columnas = "";
        String sql_parametros = "";
        for (Columna columna : this.listaColumnas) {
            if (!columna.getEsAutoGenerado()) {
                if (!sql_columnas.isBlank()) {
                    sql_columnas = sql_columnas.concat(", ");
                    sql_parametros = sql_parametros.concat(", ");
                }
                sql_columnas = sql_columnas.concat(columna.getNombre());
                sql_parametros = sql_parametros.concat("?");
            }
        }
        sql = sql.concat(sql_columnas);
        sql = sql.concat(") VALUES (");
        sql = sql.concat(sql_parametros);
        sql = sql.concat(")");
        return sql;
    }

    protected String generarSQLParaModificacion() {
        //sentencia SQL a generar es similar a 
        //UPDATE INV_ALMACENES SET NOMBRE=?, ALMACEN_CENTRAL=? WHERE ALMACEN_ID=?
        String sql = "UPDATE ";
        sql = sql.concat(this.nombre_tabla);
        sql = sql.concat(" SET ");
        String sql_columnas = "";
        String sql_predicado = "";
        for (Columna columna : this.listaColumnas) {
            if (columna.getEsllavePrimaria()) {
                if (!sql_predicado.isBlank()) {
                    sql_predicado = sql_predicado.concat(", ");
                }
                sql_predicado = sql_predicado.concat(columna.getNombre());
                sql_predicado = sql_predicado.concat("=?");
            } else {
                if (!sql_columnas.isBlank()) {
                    sql_columnas = sql_columnas.concat(", ");
                }
                sql_columnas = sql_columnas.concat(columna.getNombre());
                sql_columnas = sql_columnas.concat("=?");
            }
        }
        sql = sql.concat(sql_columnas);
        sql = sql.concat(" WHERE ");
        sql = sql.concat(sql_predicado);
        return sql;
    }

    protected String generarSQLParaEliminacion() {
        //sentencia SQL a generar es similar a 
        //DELETE FROM INV_ALMACENES WHERE ALMACEN_ID=?
        String sql = "DELETE FROM ";
        sql = sql.concat(this.nombre_tabla);
        sql = sql.concat(" WHERE ");
        String sql_predicado = "";
        for (Columna columna : this.listaColumnas) {
            if (columna.getEsllavePrimaria()) {
                if (!sql_predicado.isBlank()) {
                    sql_predicado = sql_predicado.concat(", ");
                }
                sql_predicado = sql_predicado.concat(columna.getNombre());
                sql_predicado = sql_predicado.concat("=?");
            }
        }
        sql = sql.concat(sql_predicado);
        return sql;
    }

    protected String generarSQLParaObtenerPorId() {
        //sentencia SQL a generar es similar a 
        //SELECT ALMACEN_ID, NOMBRE, ALMACEN_CENTRAL FROM INV_ALMACENES WHERE ALMACEN_ID = ?
        String sql = "SELECT ";
        String sql_columnas = "";
        String sql_predicado = "";
        for (Columna columna : this.listaColumnas) {
            if (columna.getEsllavePrimaria()) {
                if (!sql_predicado.isBlank()) {
                    sql_predicado = sql_predicado.concat(", ");
                }
                sql_predicado = sql_predicado.concat(columna.getNombre());
                sql_predicado = sql_predicado.concat("=?");
            }
            if (!sql_columnas.isBlank()) {
                sql_columnas = sql_columnas.concat(", ");
            }
            sql_columnas = sql_columnas.concat(columna.getNombre());
        }
        sql = sql.concat(sql_columnas);
        sql = sql.concat(" FROM ");
        sql = sql.concat(this.nombre_tabla);
        sql = sql.concat(" WHERE ");
        sql = sql.concat(sql_predicado);
        return sql;
    }

    protected String generarSQLParaListarTodos() {
        //sentencia SQL a generar es similar a 
        //SELECT ALMACEN_ID, NOMBRE, ALMACEN_CENTRAL FROM INV_ALMACENES
        String sql = "SELECT ";
        String sql_columnas = "";
        for (Columna columna : this.listaColumnas) {
            if (!sql_columnas.isBlank()) {
                sql_columnas = sql_columnas.concat(", ");
            }
            sql_columnas = sql_columnas.concat(columna.getNombre());
        }
        sql = sql.concat(sql_columnas);
        sql = sql.concat(" FROM ");
        sql = sql.concat(this.nombre_tabla);
        return sql;
    }

    protected void incluirValorDeParametrosParaInsercion() throws SQLException {
        throw new UnsupportedOperationException("Método no sobreescrito.");
    }

    protected void incluirValorDeParametrosParaModificacion() throws SQLException {
        throw new UnsupportedOperationException("Método no sobreescrito.");
    }

    protected void incluirValorDeParametrosParaEliminacion() throws SQLException {
        throw new UnsupportedOperationException("Método no sobreescrito.");
    }

    public Integer retornarUltimoAutoGenerado() {
        Integer resultado = null;
        try {
            String sql = DBManager.getInstance().retornarSQLParaUltimoAutoGenerado();
            this.statement = this.conexion.prepareCall(sql);
            this.resultSet = this.statement.executeQuery();
            if (this.resultSet.next()) {
                resultado = this.resultSet.getInt("id");
            }
        } catch (SQLException ex) {
            System.err.println("Error al intentar retornarUltimoAutoGenerado - " + ex);
        }
        return resultado;
    }

    public void obtenerPorId() {
        try {
            this.abrirConexion();
            String sql = this.generarSQLParaObtenerPorId();
            this.colocarSQLenStatement(sql);
            this.incluirValorDeParametrosParaObtenerPorId();
            this.ejecutarConsultaEnBD();
            if (this.resultSet.next()) {
                instanciarObjetoDelResultSet();
            } else {
                limpiarObjetoDelResultSet();
            }
        } catch (SQLException ex) {
            System.err.println("Error al intentar obtenerPorId - " + ex);
        } finally {
            try {
                this.cerrarConexion();
            } catch (SQLException ex) {
                System.err.println("Error al cerrar la conexión - " + ex);
            }
        }
    }

    public List listarTodos() {
        String sql = null;
        Consumer inclurValorDeParametro = null;
        Object parametros = null;
        return this.listarTodos(sql, inclurValorDeParametro, parametros);
    }

    public List listarTodos(String sql, Consumer inclurValorDeParametro, Object parametros) {
        List lista = new ArrayList<>();
        try {
            this.abrirConexion();
            if (sql == null) {
                sql = this.generarSQLParaListarTodos();
            }
            this.colocarSQLenStatement(sql);
            if (inclurValorDeParametro != null) {
                inclurValorDeParametro.accept(parametros);
            }
            this.ejecutarConsultaEnBD();
            while (this.resultSet.next()) {
                agregarObjetoALaLista(lista);
            }
        } catch (SQLException ex) {
            System.err.println("Error al intentar listarTodos - " + ex);
        } finally {
            try {
                this.cerrarConexion();
            } catch (SQLException ex) {
                System.err.println("Error al cerrar la conexión - " + ex);
            }
        }
        return lista;
    }

    protected void incluirValorDeParametrosParaObtenerPorId() throws SQLException {
        throw new UnsupportedOperationException("Método no sobreescrito.");
    }

    protected void instanciarObjetoDelResultSet() throws SQLException {
        throw new UnsupportedOperationException("Método no sobreescrito.");
    }

    protected void limpiarObjetoDelResultSet() {
        throw new UnsupportedOperationException("Método no sobreescrito.");
    }

    protected void agregarObjetoALaLista(List lista) throws SQLException {
        throw new UnsupportedOperationException("Método no sobreescrito.");
    }

    protected void ejecutarProcedimientoAlmacenado(String sql, Boolean conTransaccion) {
        Consumer incluirValorDeParametros = null;
        Object parametros = null;
        this.ejecutarProcedimientoAlmacenado(sql, incluirValorDeParametros, parametros, conTransaccion);
    }

    protected void ejecutarProcedimientoAlmacenado(String sql, Consumer incluirValorDeParametros, Object parametros, Boolean conTransaccion) {
        try {
            if (conTransaccion) {
                this.iniciarTransaccion();
            } else {
                this.abrirConexion();
            }
            this.colocarSQLenStatement(sql);
            if (incluirValorDeParametros != null) {
                incluirValorDeParametros.accept(parametros);
            }
            this.ejecutarModificacionEnBD();
            if (conTransaccion) {
                this.comitarTransaccion();
            }
        } catch (SQLException ex) {
            if (conTransaccion) {
                try {
                    this.rollbackTransaccion();
                } catch (SQLException ex1) {
                    Logger.getLogger(DAOImplBase.class.getName()).log(Level.SEVERE, null, ex1);
                }
            }
            Logger.getLogger(DAOImplBase.class.getName()).log(Level.SEVERE, null, ex);
        } finally {
            try {
                this.cerrarConexion();
            } catch (SQLException ex) {
                Logger.getLogger(DAOImplBase.class.getName()).log(Level.SEVERE, null, ex);
            }
        }

    }
}

File: 0682/Sesión 11 - backend en C#/SoftInv/SoftInvPersistance/DAOImpl/DAOImplBase.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data.Common;
using System.Data.SqlClient;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using SoftInv.DAOImpl.Util;
using SoftInv.Db;
using SoftInv.Model;

namespace SoftInv.DAOImpl
{
    public abstract class DAOImplBase
    {
        protected string nombreDeTabla;
        protected BindingList<Columna> listaColumnas;
        protected DbConnection conexion;
        protected DbTransaction transaccion;
        protected DbCommand comando;
        protected DbDataReader lector;
        protected bool mostrarSentenciaSQL;
        protected bool retornarLlavePrimaria;
        protected bool usarTransaccion;

        public DAOImplBase(String nombreDeTabla)
        {
            this.nombreDeTabla = nombreDeTabla;
            this.retornarLlavePrimaria = false;
            this.mostrarSentenciaSQL = true;
            this.usarTransaccion = true;
            this.IncluirListaDeColumnas();
        }

        private void IncluirListaDeColumnas()
        {
            this.listaColumnas = new BindingList<Columna>();
            this.ConfigurarListaDeColumnas();
        }

        protected abstract void ConfigurarListaDeColumnas();

        protected void AbrirConexion()
        {
            this.conexion = DBManager.Instance.Connection;
            this.conexion.Open();
        }

        protected void CerrarConexion()
        {
            if (this.conexion != null)
            {
                this.conexion.Close();
            }
        }

        protected void IniciarTransaccion()
        {
            this.AbrirConexion();
            this.transaccion = this.conexion.BeginTransaction();
        }

        protected void ComitarTransaccion()
        {
            this.transaccion.Commit();
            this.transaccion = null;
        }

        protected void RollbackTransaccion()
        {
            if (this.transaccion != null)
            {
                this.transaccion.Rollback();
            }
            this.transaccion = null;
        }

        protected void ColocarSQLenComando(string sql)
        {
            this.comando.Connection = this.conexion;
            this.comando.CommandText = sql;
            this.comando.CommandType = System.Data.CommandType.Text;
        }

        protected void EjecutarConsultaEnBD()
        {
            this.lector = this.comando.ExecuteReader();
        }

        protected int Insertar()
        {
            return this.Ejecuta_DML(TipoOperacionDML.INSERTAR);
        }

        protected int Modificar()
        {
            return this.Ejecuta_DML(TipoOperacionDML.MODIFICAR);
        }

        protected int Eliminar()
        {
            return this.Ejecuta_DML(TipoOperacionDML.ELIMINAR);
        }

        private int Ejecuta_DML(TipoOperacionDML tipoOperacion)
        {
            int resultado = 0;
            try
            {
                if (this.usarTransaccion)
                {
                    this.IniciarTransaccion();
                }
                else
                {
                    this.AbrirConexion();
                }
                using (this.comando = DBManager.Instance.CrearComando()) ;
                string sql = "";
                switch (tipoOperacion)
                {
                    case TipoOperacionDML.INSERTAR:
                        sql = this.GenerarSQLParaInsercion();
                        break;
                    case TipoOperacionDML.MODIFICAR:
                        sql = this.GenerarSQLParaModificacion();
                        break;
                    case TipoOperacionDML.ELIMINAR:
                        sql = this.GenerarSQLParaEliminacion();
                        break;
                }
                this.ColocarSQLenComando(sql);
                switch (tipoOperacion)
                {
                    case TipoOperacionDML.INSERTAR:
                        this.IncluirValorDeParametrosParaInsercion();
                        break;
                    case TipoOperacionDML.MODIFICAR:
                        this.IncluirValorDeParametrosParaModificacion();
                        break;
                    case TipoOperacionDML.ELIMINAR:
                        this.IncluirValorDeParametrosParaEliminacion();
                        break;
                }
                this.comando.ExecuteNonQuery();
                if (this.retornarLlavePrimaria)
                {
                    int id = this.RetornarUltimoAutoGenerado();
                    resultado = id;
                }
                if (this.usarTransaccion)
                {
                    this.ComitarTransaccion();
                }
            }
            catch (Exception ex)
            {
                if (this.usarTransaccion)
                {
                    this.RollbackTransaccion();
                }
                throw new Exception(ex.Message);
            }
            finally
            {
                this.CerrarConexion();
            }
            return resultado;
        }

        protected string GenerarSQLParaInsercion()
        {
            //sentencia SQL a generar es similar a 
            //INSERT INTO INV_ALMACENES (NOMBRE, ALMACEN_CENTRAL) VALUES (?,?)
            string sql = "INSERT INTO ";
            sql += this.nombreDeTabla;
            sql += "(";
            string sql_columnas = "";
            string sql_parametros = "";
            foreach (Columna columna in this.listaColumnas)
            {
                if ((bool)!columna.EsAutoGenerado)
                {
                    if (!string.IsNullOrWhiteSpace(sql_columnas))
                    {
                        sql_columnas += ", ";
                        sql_parametros += ", ";
                    }
                    sql_columnas += columna.Nombre;
                    sql_parametros += "@" + columna.Nombre;
                }
            }
            sql += sql_columnas;
            sql += ") VALUES (";
            sql += sql_parametros;
            sql += ")";
            return sql;
        }

        protected string GenerarSQLParaModificacion()
        {
            //sentencia SQL a generar es similar a 
            //UPDATE INV_ALMACENES SET NOMBRE=?, ALMACEN_CENTRAL=? WHERE ALMACEN_ID=?
            string sql = "UPDATE ";
            sql += this.nombreDeTabla;
            sql += " SET ";
            string sql_columnas = "";
            string sql_predicado = "";
            foreach (Columna columna in this.listaColumnas)
            {
                if ((bool)columna.EsLlavePrimaria)
                {
                    if (!string.IsNullOrWhiteSpace(sql_predicado))
                    {
                        sql_predicado += ", ";
                    }
                    sql_predicado += columna.Nombre;
                    sql_predicado += "=@" + columna.Nombre;
                }
                else
                {
                    if (!string.IsNullOrWhiteSpace(sql_columnas))
                    {
                        sql_columnas += ", ";
                    }
                    sql_columnas += columna.Nombre;
                    sql_columnas += "=@" + columna.Nombre;
                }
            }
            sql += sql_columnas;
            sql += " WHERE ";
            sql += sql_predicado;
            return sql;
        }

        protected string GenerarSQLParaEliminacion()
        {
            //sentencia SQL a generar es similar a 
            //DELETE FROM INV_ALMACENES WHERE ALMACEN_ID=?
            string sql = "DELETE FROM ";
            sql += this.nombreDeTabla;
            sql += " WHERE ";
            string sql_predicado = "";
            foreach (Columna columna in this.listaColumnas)
            {
                if ((bool)columna.EsLlavePrimaria)
                {
                    if (!string.IsNullOrWhiteSpace(sql_predicado))
                    {
                        sql_predicado += ", ";
                    }
                    sql_predicado += columna.Nombre;
                    sql_predicado += "=@" + columna.Nombre;
                }
            }
            sql += sql_predicado;
            return sql;
        }

        protected string GenerarSQLParaObtenerPorId()
        {
            //sentencia SQL a generar es similar a 
            //SELECT ALMACEN_ID, NOMBRE, ALMACEN_CENTRAL FROM INV_ALMACENES WHERE ALMACEN_ID = ?
            string sql = "SELECT ";
            string sql_columnas = "";
            string sql_predicado = "";
            foreach (Columna columna in this.listaColumnas)
            {
                if ((bool)columna.EsLlavePrimaria)
                {
                    if (!string.IsNullOrWhiteSpace(sql_predicado))
                    {
                        sql_predicado += ", ";
                    }
                    sql_predicado += columna.Nombre;
                    sql_predicado += "=@" + columna.Nombre;
                }
                if (!string.IsNullOrWhiteSpace(sql_columnas))
                {
                    sql_columnas += ", ";
                }
                sql_columnas += columna.Nombre;
            }
            sql += sql_columnas;
            sql += " FROM ";
            sql += this.nombreDeTabla;
            sql += " WHERE ";
            sql += sql_predicado;
            return sql;
        }

        protected string GenerarSQLParaListarTodos()
        {
            //sentencia SQL a generar es similar a 
            //SELECT ALMACEN_ID, NOMBRE, ALMACEN_CENTRAL FROM INV_ALMACENES
            string sql = "SELECT ";
            string sql_columnas = "";
            foreach (Columna columna in this.listaColumnas)
            {
                if (!string.IsNullOrWhiteSpace(sql_columnas))
                {
                    sql_columnas += ", ";
                }
                sql_columnas += columna.Nombre;
            }
            sql += sql_columnas;
            sql += " FROM ";
            sql += this.nombreDeTabla;
            return sql;
        }
        protected virtual void IncluirValorDeParametrosParaInsercion()
        {
            throw new NotImplementedException("Método no sobreescrito.");
        }

        protected virtual void IncluirValorDeParametrosParaModificacion()
        {
            throw new NotImplementedException("Método no sobreescrito.");
        }

        protected virtual void IncluirValorDeParametrosParaEliminacion()
        {
            throw new NotImplementedException("Método no sobreescrito.");
        }

        private int RetornarUltimoAutoGenerado()
        {
            int resultado = -1;
            String sql = DBManager.Instance.RetornarSQLParaUltimoAutoGenerado();
            this.ColocarSQLenComando(sql);
            this.EjecutarConsultaEnBD();
            if (this.lector.Read())
            {
                resultado = this.lector.GetInt32(0);
            }
            this.lector.Close();
            return resultado;
        }

        protected void ObtenerPorId()
        {
            this.comando = DBManager.Instance.CrearComando();
            try
            {
                this.AbrirConexion();
                string sql = this.GenerarSQLParaObtenerPorId();
                this.IncluirValorDeParametrosParaObtenerPorId();
                this.ColocarSQLenComando(sql);
                this.EjecutarConsultaEnBD();
                if (this.lector.Read())
                    this.InstanciarObjetoDelResultSet(this.lector);
                else
                    this.LimpiarObjetoDelResultSet();
            }
            catch (Exception ex)
            {
                throw new Exception(ex.Message);
            }
            finally
            {
                this.CerrarConexion();
            }
        }

        protected virtual void IncluirValorDeParametrosParaObtenerPorId()
        {
            throw new NotImplementedException("Método no sobreescrito.");
        }

        protected virtual void LimpiarObjetoDelResultSet()
        {
            throw new NotImplementedException("Método no sobreescrito.");
        }

        protected virtual void InstanciarObjetoDelResultSet(DbDataReader lector)
        {
            throw new NotImplementedException("Método no sobreescrito.");
        }

        protected BindingList<Object> ListarTodos()
        {
            BindingList<Object> lista = new BindingList<Object>();
            this.comando = DBManager.Instance.CrearComando();
            try
            {
                this.AbrirConexion();
                string sql = this.GenerarSQLParaListarTodos();
                this.ColocarSQLenComando(sql);
                this.EjecutarConsultaEnBD();
                while (this.lector.Read())
                {
                    this.AgregarObjetoALaLista(lista, this.lector);
                }
            }
            catch (Exception ex)
            {
                throw new Exception(ex.Message);
            }
            finally
            {
                this.CerrarConexion();
            }
            return lista;
        }

        protected virtual void AgregarObjetoALaLista(BindingList<Object> lista, DbDataReader lector)
        {
            throw new NotImplementedException("Método no sobreescrito.");
        }

        protected void AgregarParametro(string nombre, object value)
        {
            DbParameter parametero = this.comando.CreateParameter();
            parametero.ParameterName = nombre;
            parametero.Value = value;
            this.comando.Parameters.Add(parametero);
        }
    }
}
